src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -5922,12 +5932,24 @@ private void ParseTypeArgumentList(out SyntaxToken open, SeparatedSyntaxListBuil
             // remaining types & commas
             while (true)
             {
-                if (this.CurrentToken.Kind == SyntaxKind.GreaterThanToken)
+                // Those token kinds are our terminator token kinds that break type argument parsing
+                // It is more useful to report fewer redundant and cryptic errors than to hint using
+                // tuples as possible types
+                if (tokenKindBreaksArgumentList(this.CurrentToken.Kind) || this.CurrentToken.Kind is SyntaxKind.GreaterThanToken)
                 {
                     break;
                 }
                 else if (this.CurrentToken.Kind == SyntaxKind.CommaToken || this.IsPossibleType())
                 {
+                    if (this.CurrentToken.Kind is SyntaxKind.IdentifierToken)
+                    {
+                        if (tokenKindBreaksArgumentList(this.PeekToken(1).Kind))
+                        {
+                            // do not eat the current token as a possible type argument/param
```

Comment:
nit: consider consolidating the `if` statements
nit: consider adding an example/comment for this case